HORNETQ-1551(2) Consumer hangs on corrupted 'forced delivery' message

gaohoward opened the pull request:
It can happen when the consumer sends a 'forced delivery' then
waits forever while the connection is broken and the server's
'forced delivery' message never gets to consumer. If session is
reconnected, its consumer never knows and stays waiting.

To fix this, send a "forced delivery" message as part of session
reconnection.

clebertsuconic commented on hornetq-core-client/src/main/java/org/hornetq/core/client/impl/ClientConsumerImpl.java:
```diff
@@ -132,7 +132,7 @@
    private boolean stopped = false;
 
-   private long forceDeliveryCount;
+   private volatile long forceDeliveryCount;
```
why make it volatile?

gaohoward replied:
I make this volatile because it can be accessed by a different thread (like in this case the connection failover thread calling getForceDeliveryCount()).

gaohoward replied:
@clebertsuconic ok done. Thanks Clebert!

gaohoward replied:
doh! I made a stupid mistake (using & instead of &&). Let me correct it. :)

gaohoward replied:
corrected.

clebertsuconic commented on hornetq-core-client/src/main/java/org/hornetq/core/client/impl/ClientSessionImpl.java:
```diff
@@ -1174,6 +1174,20 @@ public void handleFailover(final CoreRemotingConnection backupConnection, Hornet
+                     //(in which case the consumer will wake up, thus avoid the infinite waiting).
+                     if (started)
+                     {
+                        SessionForceConsumerDelivery forceDel = new SessionForceConsumerDelivery(entry.getKey(), entry.getValue().getForceDeliveryCount() - 1);
```
awesome fix.. we need it in artemis!

clebertsuconic replied:
There's a bug here though.. you need to check if getForceDeliveryCount() > 0 before calling this.


I would make it:


```java
if (started && entry.getValue().getForceDeliveryCount() > 0) {
```

gaohoward replied:
ok I'll fix this (as well as in artemis)

howardgao committed:
HORNETQ-1551(2) Consumer hangs on corrupted 'forced delivery' message

It can happen when the consumer sends a 'forced delivery' then
waits forever while the connection is broken and the server's
'forced delivery' message never gets to consumer. If session is
reconnected, its consumer never knows and stays waiting.

To fix this, send a "forced delivery" message as part of session
reconnection.

clebertsuconic commented:
merged manually... closing
